src/devices/Mcp3xxx/Mcp3xxx.cs

```diff
@@ -0,0 +1,158 @@
+        /// <param name="valueChannel">Channel which represents the signal driving the value in a positive direction (valid values: 0 to channelcount - 1).</param>
+        /// <param name="referenceChannel">Channel which represents the signal driving the value in a negative direction (valid values: 0 to channelcount - 1).</param>
+        /// <returns>A value corresponding to relative voltage level on specified device channels</returns>
+        public virtual int ReadDifferential(int valueChannel, int referenceChannel)
+        {
+            CheckChannelRange(valueChannel, ChannelCount);
+            CheckChannelRange(referenceChannel, ChannelCount);
+
+            if (valueChannel == referenceChannel)
+            {
+                throw new ArgumentException($"ADC differential channels must be different.", nameof(valueChannel) + " " + nameof(referenceChannel));
+            }
+
+            return ReadInternal(valueChannel, InputType.SingleEnded, _adcResolutionBits) -
+                   ReadInternal(referenceChannel, InputType.SingleEnded, _adcResolutionBits);
+        }
+
+        /// <summary>
+        /// Reads a value from the device
+        /// </summary>
+        /// <param name="channel">Channel which value should be read from (valid values: 0 to channelcount - 1)</param>
+        /// <returns>A value corresponding to relative voltage level on specified device channel</returns>
+        public virtual int Read(int channel)
+        {
+            return ReadInternal(channel, InputType.SingleEnded, _adcResolutionBits);
+        }
+
+        /// <summary>
+        /// Reads a value from the device
+        /// </summary>
+        /// <param name="channel">Channel which value should be read from. For diffential inputs this represents a channel pair (valid values: 0 - channelcount - 1 or 0 - channelcount / 2 - 1  with differential inputs)</param>
+        /// <param name="inputType">The type of input channel to read.</param>
+        /// <param name="adcResolutionBits">The number of bits in the returned value</param>
+        /// <returns>A value corresponding to relative voltage level on specified device channel</returns>
+        protected int ReadInternal(int channel, InputType inputType, int adcResolutionBits)
+        {
+            int channelVal;
+            int requestVal;
+
+            CheckChannelRange(channel, inputType == InputType.SingleEnded ? ChannelCount : ChannelCount / 2);
+
+            // create a value that represents the channel value. For differental inputs
+            // then incorporate the lower bit which indicates if the channel is inverted or not.
+            channelVal = inputType switch
+            {
+                InputType.Differential => channel * 2,
+                InputType.InvertedDifferential => channel * 2 + 1,
+                _ => channel
+            };
+
+            // create a value to represent the request to the ADC 
+            requestVal = ChannelCount switch
```

Comment:
please use regular switch syntax, while I like functional programming this is inconsistent with the rest of the code base.